test_json.py fails

Content of json file located in user-temp directory:
```
[{"amount": 100, "name": "Alice"}, {"amount": 200, "name": "Bob"}]
```


```
$ py.test test_json.py
←[1m============================= test session starts =============================←[0m
platform win32 -- Python 2.7.8 -- py-1.4.26 -- pytest-2.6.4
collected 21 items ←[0m←[1m
←[0m
test_json.py ...FF........F....F..

================================== FAILURES ===================================
________________________________ test_resource ________________________________

←[1m    def test_resource():←[0m
←[1m>       assert isinstance(resource('jsonlines://foo.json'), JSONLines)←[0m
←[1m←[31mE       assert isinstance(<into.directory._Directory object at 0x0422CEB0>, JSONLines)←[0m
←[1m←[31mE        +  where <into.directory._Directory object at 0x0422CEB0> = resource('jsonlines://foo.json')←[0m

test_json.py:55: AssertionError
___________________________ test_resource_guessing ____________________________

←[1m    def test_resource_guessing():←[0m
←[1m        with json_file(dat) as fn:←[0m
←[1m>           assert isinstance(resource(fn), JSON)←[0m

test_json.py:63:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
..\..\regex.py:64: in __call__
←[1m    return self.dispatch(s)(s, *args, **kwargs)←[0m
..\..\directory.py:57: in resource_directory
←[1m    subtype = type(resource(one_uri, **kwargs))←[0m
..\..\regex.py:64: in __call__
←[1m    return self.dispatch(s)(s, *args, **kwargs)←[0m
!!! Recursion detected (same locals & position)
___________________________ test_multiple_jsonlines ___________________________

←[1m    def test_multiple_jsonlines():←[0m
←[1m        try:←[0m
←[1m            with open('_test_a1.json', 'w') as f:←[0m
←[1m                json.dump(dat, f)←[0m
←[1m            with open('_test_a2.json', 'w') as f:←[0m
←[1m                json.dump(dat, f)←[0m
←[1m>           r = resource('_test_a*.json')←[0m

test_json.py:154:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
..\..\regex.py:64: in __call__
←[1m    return self.dispatch(s)(s, *args, **kwargs)←[0m
..\..\directory.py:57: in resource_directory
←[1m    subtype = type(resource(one_uri, **kwargs))←[0m
..\..\regex.py:64: in __call__
←[1m    return self.dispatch(s)(s, *args, **kwargs)←[0m
..\..\directory.py:57: in resource_directory
←[1m    subtype = type(resource(one_uri, **kwargs))←[0m
..\..\regex.py:64: in __call__
←[1m    return self.dispatch(s)(s, *args, **kwargs)←[0m
!!! Recursion detected (same locals & position)
_____________________________ test_resource_gzip ______________________________

←[1m    def test_resource_gzip():←[0m
←[1m>       assert isinstance(resource('foo.json.gz'), (JSON, JSONLines))←[0m
←[1m←[31mE       assert isinstance(<into.directory._Directory object at 0x04364030>, (<class 'into.backends.json.JSON'>, <cla
ss 'into.backends.json.JSONLines'>))←[0m
←[1m←[31mE        +  where <into.directory._Directory object at 0x04364030> = resource('foo.json.gz')←[0m

test_json.py:206: AssertionError
←[1m←[31m===================== 4 failed, 17 passed in 2.78 seconds =====================←[0m```

--- a/into/backends/h5py.py
+++ b/into/backends/h5py.py
@@ -252,6 +252,7 @@ def drop(h):
 
 @dispatch(h5py.File)
 def drop(h):
+    h.close()
     os.remove(h.filename)
 
 

--- a/into/backends/h5py.py
+++ b/into/backends/h5py.py
@@ -252,8 +252,9 @@ def drop(h):
 
 @dispatch(h5py.File)
 def drop(h):
+    fn = h.filename
     h.close()
-    os.remove(h.filename)
+    os.remove(fn)
 
 
 ooc_types.add(h5py.Dataset)

--- a/into/backends/sql_csv.py
+++ b/into/backends/sql_csv.py
@@ -98,6 +98,7 @@ def execute_copy_all(dialect, engine, statement):
     cursor = conn.cursor()
     cursor.execute(statement)
     conn.commit()
+    conn.close()
 
 
 @append.register(sqlalchemy.Table, CSV)

--- a/into/backends/sql_csv.py
+++ b/into/backends/sql_csv.py
@@ -64,7 +64,7 @@ def copy_mysql(dialect, tbl, csv):
     delimiter = csv.dialect.get('delimiter', ',')
     quotechar = csv.dialect.get('quotechar', '"')
     escapechar = csv.dialect.get('escapechar', '\\')
-    lineterminator = csv.dialect.get('lineterminator', r'\n\r')
+    lineterminator = csv.dialect.get('lineterminator', '\\n\\r')
     skiprows = 1 if csv.has_header else 0
     encoding = csv.encoding or 'utf-8'
 

--- a/into/backends/sql_csv.py
+++ b/into/backends/sql_csv.py
@@ -13,6 +13,8 @@ execute_copy = RegexDispatcher('execute_copy')
 @copy_command.register('.*sqlite')
 def copy_sqlite(dialect, tbl, csv):
     abspath = os.path.abspath(csv.path)
+    if os.name == 'nt':
+        abspath = abspath.replace('\\', '\\\\')
     tblname = tbl.name
     dbpath = str(tbl.bind.url).split('///')[-1]
 
@@ -33,6 +35,8 @@ def execute_copy_sqlite(dialect, engine, statement):
 @copy_command.register('postgresql')
 def copy_postgres(dialect, tbl, csv):
     abspath = os.path.abspath(csv.path)
+    if os.name == 'nt':
+        abspath = abspath.replace('\\', '\\\\')
     tblname = tbl.name
     format_str = 'csv'
     delimiter = csv.dialect.get('delimiter', ',')
@@ -60,6 +64,8 @@ def copy_postgres(dialect, tbl, csv):
 def copy_mysql(dialect, tbl, csv):
     mysql_local = ''
     abspath = os.path.abspath(csv.path)
+    if os.name == 'nt':
+        abspath = abspath.replace('\\', '\\\\')
     tblname = tbl.name
     delimiter = csv.dialect.get('delimiter', ',')
     quotechar = csv.dialect.get('quotechar', '"')

--- a/into/backends/sql_csv.py
+++ b/into/backends/sql_csv.py
@@ -2,6 +2,7 @@
 from ..regex import RegexDispatcher
 from ..append import append
 from .csv import CSV
+from multipledispatch import MDNotImplementedError
 import os
 import datashape
 import sqlalchemy
@@ -27,6 +28,8 @@ def copy_sqlite(dialect, tbl, csv):
 
 @execute_copy.register('sqlite')
 def execute_copy_sqlite(dialect, engine, statement):
+    if os.name == 'nt':
+        raise MDNotImplementedError()
     ps = subprocess.Popen(statement, shell=True, stdout=subprocess.PIPE)
     return ps.stdout.read()
 

--- a/into/backends/tests/test_bcolz.py
+++ b/into/backends/tests/test_bcolz.py
@@ -5,7 +5,7 @@ from into.backends.bcolz import (append, convert, ctable, carray, resource,
 from into.chunks import chunks
 from into import append, convert, discover, into
 import numpy as np
-from into.utils import tmpfile
+from into.utils import tmpfile, ignoring
 from contextlib import contextmanager
 import shutil
 import os
@@ -20,6 +20,8 @@ def tmpbcolz(*args, **kwargs):
     try:
         yield r
     finally:
+        with ignoring(Exception):
+            r.flush()
         if os.path.exists(fn):
             shutil.rmtree(fn)
 

--- a/into/backends/tests/test_bcolz.py
+++ b/into/backends/tests/test_bcolz.py
@@ -6,7 +6,23 @@ from into.chunks import chunks
 from into import append, convert, discover, into
 import numpy as np
 from into.utils import tmpfile
+from contextlib import contextmanager
+import shutil
 import os
+import uuid
+
+
+@contextmanager
+def tmpbcolz(*args, **kwargs):
+    fn = '.%s.bcolz' % str(uuid.uuid1())
+    r = resource(fn, *args, **kwargs)
+
+    try:
+        yield r
+    finally:
+        if os.path.exists(fn):
+            shutil.rmtree(fn)
+
 
 
 def eq(a, b):
@@ -52,10 +68,7 @@ def test_append_other():
 
 
 def test_resource_ctable():
-    with tmpfile('.bcolz') as fn:
-        r = resource(fn,
-                     dshape='var * {name: string[5, "ascii"], balance: int32}')
-
+    with tmpbcolz(dshape='var * {name: string[5, "ascii"], balance: int32}') as r:
         assert isinstance(r, ctable)
         assert r.dtype == [('name', 'S5'), ('balance', 'i4')]
 
@@ -66,48 +79,37 @@ def get_expectedlen(x):
 
 
 def test_resource_ctable_overrides_expectedlen():
-    with tmpfile('.bcolz') as fn:
-        r = resource(fn,
-                     dshape='100 * {name: string[5, "ascii"], balance: int32}',
-                     expectedlen=200)
-
+    with tmpbcolz(dshape='100 * {name: string[5, "ascii"], balance: int32}',
+                  expectedlen=200) as r:
         assert isinstance(r, ctable)
         assert r.dtype == [('name', 'S5'), ('balance', 'i4')]
         assert all(get_expectedlen(r[c]) == 200 for c in r.names)
 
 
 def test_resource_ctable_correctly_infers_length():
-    with tmpfile('.bcolz') as fn:
-        r = resource(fn,
-                     dshape='100 * {name: string[5, "ascii"], balance: int32}')
-
+    with tmpbcolz(dshape='100 * {name: string[5, "ascii"], balance: int32}') as r:
         assert isinstance(r, ctable)
         assert r.dtype == [('name', 'S5'), ('balance', 'i4')]
         assert all(get_expectedlen(r[c]) == 100 for c in r.names)
 
 
 def test_resource_carray():
-    with tmpfile('.bcolz') as fn:
-        r = resource(fn, dshape='var * int32')
-
+    with tmpbcolz(dshape='var * int32') as r:
         assert isinstance(r, carray)
         assert r.dtype == 'i4'
         assert r.shape == (0,)
 
 
 def test_resource_existing_carray():
-    with tmpfile('.bcolz') as fn:
-        r = resource(fn, dshape='var * int32')
+    with tmpbcolz(dshape='var * int32') as r:
         append(r, [1, 2, 3])
         r.flush()
-        newr = resource(fn)
+        newr = resource(r.rootdir)
         assert isinstance(newr, carray)
 
 
 def test_resource_carray_overrides_expectedlen():
-    with tmpfile('.bcolz') as fn:
-        r = resource(fn, dshape='100 * int32', expectedlen=200)
-
+    with tmpbcolz(dshape='100 * int32', expectedlen=200) as r:
         assert isinstance(r, carray)
         assert r.dtype == 'i4'
         assert r.shape == (100,)
@@ -115,9 +117,7 @@ def test_resource_carray_overrides_expectedlen():
 
 
 def test_resource_ctable_correctly_infers_length():
-    with tmpfile('.bcolz') as fn:
-        r = resource(fn, dshape='100 * int32')
-
+    with tmpbcolz(dshape='100 * int32') as r:
         assert isinstance(r, carray)
         assert r.dtype == 'i4'
         assert get_expectedlen(r) == 100
@@ -128,12 +128,11 @@ def test_into_respects_expected_len_during_append():
         b = into(fn, [1, 2, 3])
         assert get_expectedlen(b) == 3
         assert len(b) == 3
+        shutil.rmtree(fn)
 
 
 def test_resource_nd_carray():
-    with tmpfile('.bcolz') as fn:
-        r = resource(fn, dshape='10 * 10 * 10 * int32')
-
+    with tmpbcolz(dshape='10 * 10 * 10 * int32') as r:
         assert isinstance(r, carray)
         assert r.dtype == 'i4'
         assert r.shape == (10, 10, 10)
@@ -156,20 +155,20 @@ def test_resource_existing_ctable():
         r2 = resource(fn)
         assert eq(r2[:], y)
 
+        shutil.rmtree(fn)
+
 
 def test_drop():
-    with tmpfile('.bcolz') as fn:
-        r = resource(fn, dshape='var * {name: string[5, "ascii"], balance: int32}')
-
-        assert os.path.exists(fn)
-        drop(fn)
-        assert not os.path.exists(fn)
+    with tmpbcolz(dshape='var * {name: string[5, "ascii"], balance: int32}') as b:
+        assert os.path.exists(b.rootdir)
+        drop(b)
+        assert not os.path.exists(b.rootdir)
 
 
 def test_resource_shape():
-    with tmpfile('.bcolz') as fn:
-        assert resource(fn, dshape='10 * int').shape == (10,)
-    with tmpfile('.bcolz') as fn:
-        assert resource(fn, dshape='10 * 10 * int').shape == (10, 10)
-    with tmpfile('.bcolz') as fn:
-        assert resource(fn, dshape='var * 10 * int').shape == (0, 10)
+    with tmpbcolz(dshape='10 * int') as b:
+        assert b.shape == (10,)
+    with tmpbcolz(dshape='10 * 10 * int') as b:
+        assert b.shape == (10, 10)
+    with tmpbcolz(dshape='var * 10 * int') as b:
+        assert b.shape == (0, 10)

--- a/into/backends/tests/test_csv.py
+++ b/into/backends/tests/test_csv.py
@@ -68,6 +68,7 @@ def test_pandas_read_supports_missing_integers():
         assert df.dtypes['val'] == 'f4'
 
 
+@skipif(os.name == 'nt')
 def test_pandas_read_supports_gzip():
     with filetext('Alice,1\nBob,2', open=gzip.open,
                   mode='wt', extension='.csv.gz') as fn:
@@ -141,6 +142,7 @@ def test_pandas_loads_in_datetimes_naively():
         assert df.dtypes['when'] == 'M8[ns]'
 
 
+@skipif(os.name == 'nt')
 def test_pandas_discover_on_gzipped_files():
     with filetext('name,when\nAlice,2014-01-01\nBob,2014-02-02',
                   open=gzip.open, mode='wt', extension='.csv.gz') as fn:

--- a/into/backends/tests/test_csv.py
+++ b/into/backends/tests/test_csv.py
@@ -69,7 +69,8 @@ def test_pandas_read_supports_missing_integers():
 
 
 def test_pandas_read_supports_gzip():
-    with filetext('Alice,1\nBob,2', open=gzip.open, extension='.csv.gz') as fn:
+    with filetext('Alice,1\nBob,2', open=gzip.open,
+                  mode='wt', extension='.csv.gz') as fn:
         ds = datashape.dshape('var * {name: string, amount: int}')
         csv = CSV(fn)
         df = csv_to_DataFrame(csv, dshape=ds)
@@ -142,7 +143,7 @@ def test_pandas_loads_in_datetimes_naively():
 
 def test_pandas_discover_on_gzipped_files():
     with filetext('name,when\nAlice,2014-01-01\nBob,2014-02-02',
-            open=gzip.open, extension='.csv.gz') as fn:
+                  open=gzip.open, mode='wt', extension='.csv.gz') as fn:
         csv = CSV(fn, has_header=True)
         ds = datashape.dshape('var * {name: ?string, when: ?datetime}')
         assert discover(csv) == ds

--- a/into/backends/tests/test_h5py.py
+++ b/into/backends/tests/test_h5py.py
@@ -104,8 +104,11 @@ def test_append():
 def test_into_resource():
     with tmpfile('.hdf5') as fn:
         d = into(fn + '::/x', x)
-        assert d.shape == x.shape
-        assert eq(d[:], x[:])
+        try:
+            assert d.shape == x.shape
+            assert eq(d[:], x[:])
+        finally:
+            d.file.close()
 
 
 def test_numpy():
@@ -130,9 +133,12 @@ def test_create():
     with tmpfile('.hdf5') as fn:
         ds = datashape.dshape('{x: int32, y: {z: 3 * int32}}')
         f = create(h5py.File, dshape='{x: int32, y: {z: 3 * int32}}', path=fn)
-        assert isinstance(f, h5py.File)
-        assert f.filename == fn
-        assert discover(f) == ds
+        try:
+            assert isinstance(f, h5py.File)
+            assert f.filename == fn
+            assert discover(f) == ds
+        finally:
+            f.close()
 
 
 def test_create_partially_present_dataset():
@@ -143,21 +149,29 @@ def test_create_partially_present_dataset():
         ds2 = datashape.dshape('{x: int32, y: 5 * int32}')
         f2 = create(h5py.File, dshape=ds2, path=fn)
 
-        assert f.filename == f2.filename
-        assert list(f.keys()) == list(f2.keys())
-        assert f['y'].dtype == 'i4'
+        try:
+            assert f.filename == f2.filename
+            assert list(f.keys()) == list(f2.keys())
+            assert f['y'].dtype == 'i4'
+        finally:
+            f.close()
+            f2.close()
+
 
 
 def test_resource():
     with tmpfile('.hdf5') as fn:
         ds = datashape.dshape('{x: int32, y: 3 * int32}')
         r = resource(fn, dshape=ds)
-
-        assert isinstance(r, h5py.File)
-        assert discover(r) == ds
-
         r2 = resource(fn + '::/x')
-        assert isinstance(r2, h5py.Dataset)
+
+        try:
+            assert isinstance(r, h5py.File)
+            assert discover(r) == ds
+            assert isinstance(r2, h5py.Dataset)
+        finally:
+            r.close()
+            r2.file.close()
 
 
 def test_resource_with_datapath():
@@ -165,25 +179,33 @@ def test_resource_with_datapath():
         ds = datashape.dshape('3 * 4 * int32')
         r = resource(fn + '::/data', dshape=ds)
 
-        assert isinstance(r, h5py.Dataset)
-        assert discover(r) == ds
-        assert r.file.filename == fn
-        assert r.file['/data'] == r
+        try:
+            assert isinstance(r, h5py.Dataset)
+            assert discover(r) == ds
+            assert r.file.filename == fn
+            assert r.file['/data'] == r
+        finally:
+            r.file.close()
 
 
 def test_resource_with_variable_length():
     with tmpfile('.hdf5') as fn:
         ds = datashape.dshape('var * 4 * int32')
         r = resource(fn + '::/data', dshape=ds)
-
-        assert r.shape == (0, 4)
+        try:
+            assert r.shape == (0, 4)
+        finally:
+            r.file.close()
 
 
 def test_copy_with_into():
     with tmpfile('.hdf5') as fn:
         dset = into(fn + '::/data', [1, 2, 3])
-        assert dset.shape == (3,)
-        assert eq(dset[:], [1, 2, 3])
+        try:
+            assert dset.shape == (3,)
+            assert eq(dset[:], [1, 2, 3])
+        finally:
+            dset.file.close()
 
 
 def test_varlen_dtypes():
@@ -191,14 +213,22 @@ def test_varlen_dtypes():
                 dtype=[('name', 'O'), ('amount', 'i4')])
     with tmpfile('.hdf5') as fn:
         dset = into(fn + '::/data', y)
-
-        assert into(list, dset) == into(list, dset)
+        try:
+            assert into(list, dset) == into(list, dset)
+        finally:
+            dset.file.close()
 
 
 def test_resource_shape():
     with tmpfile('.hdf5') as fn:
-        assert resource(fn+'::/data', dshape='10 * int').shape == (10,)
+        r = resource(fn+'::/data', dshape='10 * int')
+        assert r.shape == (10,)
+        r.file.close()
     with tmpfile('.hdf5') as fn:
-        assert resource(fn+'::/data', dshape='10 * 10 * int').shape == (10, 10)
+        r = resource(fn+'::/data', dshape='10 * 10 * int')
+        assert r.shape == (10, 10)
+        r.file.close()
     with tmpfile('.hdf5') as fn:
-        assert resource(fn+'::/data', dshape='var * 10 * int').shape == (0, 10)
+        r = resource(fn+'::/data', dshape='var * 10 * int')
+        assert r.shape == (0, 10)
+        r.file.close()

--- a/into/backends/tests/test_h5py.py
+++ b/into/backends/tests/test_h5py.py
@@ -6,7 +6,7 @@ from contextlib import contextmanager
 from into.backends.h5py import append, create, resource, discover, convert
 from into.backends.h5py import unicode_dtype
 
-from into.utils import tmpfile
+from into.utils import tmpfile, ignoring
 from into.chunks import chunks
 from into import into, append, convert, discover, drop
 import datashape
@@ -24,7 +24,8 @@ def file(x):
         try:
             yield fn, f, data
         finally:
-            f.close()
+            with ignoring(Exception):
+                f.close()
 
 
 x = np.ones((2, 3), dtype='i4')
@@ -39,7 +40,8 @@ def test_drop_group():
             drop(f['/group'])
             assert '/group' not in f.keys()
         finally:
-            f.close()
+            with ignoring(Exception):
+                f.close()
 
 
 def test_drop_dataset():
@@ -52,7 +54,8 @@ def test_drop_dataset():
             drop(data)
             assert '/data' not in f.keys()
         finally:
-            f.close()
+            with ignoring(Exception):
+                f.close()
 
 
 def test_drop_file():
@@ -81,7 +84,8 @@ def test_discover_on_data_with_object_in_record_name():
             assert (discover(f['data']) ==
                     datashape.dshape('2 * {lrg_object: string, an_int: int64}'))
         finally:
-            f.close()
+            with ignoring(Exception):
+                f.close()
 
 
 def eq(a, b):

--- a/into/backends/tests/test_sql_csv.py
+++ b/into/backends/tests/test_sql_csv.py
@@ -2,6 +2,7 @@ from into.backends.sql_csv import *
 from into import resource, into
 import datashape
 from into.utils import tmpfile
+from into.compatibility import skipif
 import os
 
 
@@ -53,6 +54,7 @@ def test_mysql_load():
             IGNORE 1 LINES;""" % escaped_fn)
 
 
+@skipif(os.name == 'nt')
 def test_into_sqlite():
     data = [('Alice', 100), ('Bob', 200)]
     ds = datashape.dshape('var * {name: string, amount: int}')

--- a/into/backends/tests/test_sql_csv.py
+++ b/into/backends/tests/test_sql_csv.py
@@ -9,14 +9,14 @@ def normalize(s):
     return s2
 
 
-csv = CSV('/var/tmp/myfile.csv', delimiter=',', has_header=True)
+csv = CSV('myfile.csv', delimiter=',', has_header=True)
 ds = datashape.dshape('var * {name: string, amount: int}')
 tbl = resource('sqlite:///:memory:::my_table', dshape=ds)
 
 
 def test_postgres_load():
     assert normalize(copy_command('postgresql', tbl, csv)) == normalize(r"""
-    COPY my_table from '/var/tmp/myfile.csv'
+    COPY my_table from 'myfile.csv'
         (FORMAT csv,
          DELIMITER E',',
          NULL '',
@@ -29,13 +29,13 @@ def test_postgres_load():
 
 def test_sqlite_load():
     assert normalize(copy_command('sqlite', tbl, csv)) == normalize("""
-     (echo '.mode csv'; echo '.import /var/tmp/myfile.csv my_table';) | sqlite3 :memory:
+     (echo '.mode csv'; echo '.import myfile.csv my_table';) | sqlite3 :memory:
      """)
 
 
 def test_mysql_load():
     assert normalize(copy_command('mysql', tbl, csv)) == normalize(r"""
-            LOAD DATA  INFILE '/var/tmp/myfile.csv'
+            LOAD DATA  INFILE 'myfile.csv'
             INTO TABLE my_table
             CHARACTER SET utf-8
             FIELDS

--- a/into/backends/tests/test_sql_csv.py
+++ b/into/backends/tests/test_sql_csv.py
@@ -11,6 +11,10 @@ def normalize(s):
 
 
 fn = os.path.abspath('myfile.csv')
+if os.name == 'nt':
+    escaped_fn = fn.replace('\\', '\\\\')
+else:
+    escaped_fn = fn
 
 csv = CSV(fn, delimiter=',', has_header=True)
 ds = datashape.dshape('var * {name: string, amount: int}')
@@ -27,13 +31,13 @@ def test_postgres_load():
          ESCAPE '\\',
          HEADER True,
          ENCODING 'utf-8');
-    """ % fn)
+    """ % escaped_fn)
 
 
 def test_sqlite_load():
     assert normalize(copy_command('sqlite', tbl, csv)) == normalize("""
      (echo '.mode csv'; echo '.import %s my_table';) | sqlite3 :memory:
-     """ % fn)
+     """ % escaped_fn)
 
 
 def test_mysql_load():
@@ -46,7 +50,7 @@ def test_mysql_load():
                 ENCLOSED BY '"'
                 ESCAPED BY '\\'
             LINES TERMINATED BY '\\n\\r'
-            IGNORE 1 LINES;""" % fn)
+            IGNORE 1 LINES;""" % escaped_fn)
 
 
 def test_into_sqlite():

--- a/into/backends/tests/test_sql_csv.py
+++ b/into/backends/tests/test_sql_csv.py
@@ -2,6 +2,7 @@ from into.backends.sql_csv import *
 from into import resource, into
 import datashape
 from into.utils import tmpfile
+import os
 
 
 def normalize(s):
@@ -9,41 +10,43 @@ def normalize(s):
     return s2
 
 
-csv = CSV('myfile.csv', delimiter=',', has_header=True)
+fn = os.path.abspath('myfile.csv')
+
+csv = CSV(fn, delimiter=',', has_header=True)
 ds = datashape.dshape('var * {name: string, amount: int}')
 tbl = resource('sqlite:///:memory:::my_table', dshape=ds)
 
 
 def test_postgres_load():
-    assert normalize(copy_command('postgresql', tbl, csv)) == normalize(r"""
-    COPY my_table from 'myfile.csv'
+    assert normalize(copy_command('postgresql', tbl, csv)) == normalize("""
+    COPY my_table from '%s'
         (FORMAT csv,
          DELIMITER E',',
          NULL '',
          QUOTE '"',
-         ESCAPE '\',
+         ESCAPE '\\',
          HEADER True,
          ENCODING 'utf-8');
-    """)
+    """ % fn)
 
 
 def test_sqlite_load():
     assert normalize(copy_command('sqlite', tbl, csv)) == normalize("""
-     (echo '.mode csv'; echo '.import myfile.csv my_table';) | sqlite3 :memory:
-     """)
+     (echo '.mode csv'; echo '.import %s my_table';) | sqlite3 :memory:
+     """ % fn)
 
 
 def test_mysql_load():
-    assert normalize(copy_command('mysql', tbl, csv)) == normalize(r"""
-            LOAD DATA  INFILE 'myfile.csv'
+    assert normalize(copy_command('mysql', tbl, csv)) == normalize("""
+            LOAD DATA  INFILE '%s'
             INTO TABLE my_table
             CHARACTER SET utf-8
             FIELDS
                 TERMINATED BY ','
                 ENCLOSED BY '"'
-                ESCAPED BY '\'
-            LINES TERMINATED BY '\n\r'
-            IGNORE 1 LINES;""")
+                ESCAPED BY '\\'
+            LINES TERMINATED BY '\\n\\r'
+            IGNORE 1 LINES;""" % fn)
 
 
 def test_into_sqlite():

--- a/into/backends/tests/test_sqlite_into.py
+++ b/into/backends/tests/test_sqlite_into.py
@@ -37,6 +37,7 @@ def test_simple_into(csv):
         cursor.execute("SELECT name FROM sqlite_master WHERE type='table' and name='{0}';".format(tbl))
 
         sqlite_tbl_names = cursor.fetchall()
-        assert sqlite_tbl_names[0][0] == tbl
+        conn.close()
 
+        assert sqlite_tbl_names[0][0] == tbl
         assert into(list, t) == data

--- a/into/backends/tests/test_ssh.py
+++ b/into/backends/tests/test_ssh.py
@@ -12,6 +12,12 @@ from into.backends.ssh import SSH, resource, ssh_pattern, sftp, drop, connect
 from into.backends.csv import CSV
 from into import into, discover, CSV, JSONLines, JSON
 from into.temp import _Temp, Temp
+import socket
+
+try:
+    ssh = connect(hostname='localhost')
+except socket.error:
+    pytest.importorskip('does_not_exist')
 
 
 def test_resource():

--- a/into/directory.py
+++ b/into/directory.py
@@ -15,12 +15,12 @@ class _Directory(Chunks):
     parametrized Directory classes.
 
     >>> from into import CSV
-    >>> c = Directory(CSV)('path/to/data/')
+    >>> c = Directory(CSV)('path/to/data/')  # doctest: +SKIP
 
     Normal use through resource strings
 
-    >>> r = resource('path/to/data/*.csv')
-    >>> r = resource('path/to/data/')
+    >>> r = resource('path/to/data/*.csv')  # doctest: +SKIP
+    >>> r = resource('path/to/data/')  # doctest: +SKIP
 
 
     """

--- a/into/directory.py
+++ b/into/directory.py
@@ -42,12 +42,16 @@ Directory.__doc__ = Directory.__doc__
 Directory = memoize(Directory)
 
 
+re_path_sep = os.path.sep
+if re_path_sep == '\\':
+    re_path_sep = '\\\\'
+
 @discover.register(_Directory)
 def discover_Directory(c, **kwargs):
     return var * discover(first(c)).subshape[0]
 
 
-@resource.register('.+' + os.path.sep + '\*\..+', priority=15)
+@resource.register('.+' + re_path_sep + '\*\..+', priority=15)
 def resource_directory(uri, **kwargs):
     path = uri.rsplit(os.path.sep, 1)[0]
     try:
@@ -58,7 +62,7 @@ def resource_directory(uri, **kwargs):
     return Directory(subtype)(path, **kwargs)
 
 
-@resource.register('.+' + os.path.sep, priority=9)
+@resource.register('.+' + re_path_sep, priority=9)
 def resource_directory_with_trailing_slash(uri, **kwargs):
     try:
         one_uri = os.listdir(uri)[0]

--- a/into/tests/test_directory.py
+++ b/into/tests/test_directory.py
@@ -43,4 +43,6 @@ def test_resource_directory():
 
 
 def test_resource_directory():
-    assert isinstance(resource('/a/directory/that/doesnt/exist/'), _Directory)
+    assert isinstance(resource(os.path.join('a', 'nonexistent', 'directory') +
+                               os.path.sep),
+                      _Directory)

--- a/into/utils.py
+++ b/into/utils.py
@@ -80,7 +80,7 @@ def cls_name(cls):
 
 
 @contextmanager
-def filetext(text, extension='', open=open, mode='wt'):
+def filetext(text, extension='', open=open, mode='w'):
     with tmpfile(extension=extension) as filename:
         f = open(filename, mode=mode)
         try:

--- a/into/utils.py
+++ b/into/utils.py
@@ -46,16 +46,8 @@ def tmpfile(extension=''):
 
     yield filename
 
-    try:
-        if os.path.exists(filename):
-            os.remove(filename)
-    except OSError:  # Sometimes Windows can't close files
-        if os.name == 'nt':
-            os.close(handle)
-            try:
-                os.remove(filename)
-            except OSError:  # finally give up
-                pass
+    if os.path.exists(filename):
+        os.remove(filename)
 
 
 def keywords(func):

